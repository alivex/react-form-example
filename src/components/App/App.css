.App {
  text-align: left;
  padding: 1em;
  margin: 1em;
  border: 2px solid #d3d3d3;
  border-radius: .5em;
  vertical-align: middle;
  margin-left: auto;
  margin-right: auto;
  width: 90%;
  margin-top: 100px;
  color: black;
}

.form {
  padding: 1em;
}

.submitBtn {
  padding: 0.5rem 1rem;
  font-size: 1.25rem;
  line-height: 1.5;
  border-radius: 0.3rem;
  color: #fff;
  background-color: #007bff;
  border-color: #007bff;
  display: block;
  width: 100%;
}
