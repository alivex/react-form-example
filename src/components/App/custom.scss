$grid-breakpoints: (
  sm: 601px
);

$container-max-widths: (
  sm: 565px
);

@import '~bootstrap/scss/bootstrap';