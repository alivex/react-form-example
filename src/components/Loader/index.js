import * as React from 'react';

import { Modal, ModalBody, Spinner} from 'reactstrap';        

class Loader extends React.Component{

  render(){
      return(
        <Modal isOpen={ this.props.open } centered={true}>
          <ModalBody>
              <Spinner type="grow" color="primary" />
              <Spinner color="dark" />
              <Spinner type="grow" color="secondary" />
              <Spinner color="info" />
              <Spinner type="grow" color="warning" />
              <Spinner color="success" />
              <Spinner type="grow" color="danger" />
              <Spinner color="success" />
              <Spinner type="grow" color="danger" />
              <Spinner color="warning" />
              <Spinner type="grow" color="info" />
              <Spinner color="secondary" />
              <Spinner type="grow" color="dark" />
              <Spinner color="primary" />
          </ModalBody>
        </Modal>
      )
  }
}

export default Loader;
