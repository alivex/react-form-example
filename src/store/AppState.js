export default {
    form: [{
        name: 'nameInp',
        label:'Name',    
        type: 'text',        
        placeholder:'Enter your name'
      },
      {
        name: 'emailInp',
        label:'EMail',
        type: 'email',
        placeholder:'Enter your EMail'
      },
      {        
        name: 'messageInp',
        label:'Message',
        type: 'textarea',
        placeholder:'Enter your message'
      }
    ],
    formDisable:false,
    fields:{}
}