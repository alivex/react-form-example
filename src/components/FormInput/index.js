import * as React from 'react';

import { Input } from 'reactstrap';
import {connect} from 'react-redux';

import {updateForm} from '../../store/actions'

class FormField extends React.Component{

    handleChange=(event)=>{
      
      event.preventDefault();
      
      let fields = this.props.fieldsAsProps;
      let name = event.target.name; // Field name
      let value = event.target.value; // Field value

      fields[name] = value;
      this.props.updateForm(fields);
    }
    
    render(){
        const fields = this.props.data;
        return(
          <Input name={ fields['name'] } type={fields['type']} placeholder={fields['placeholder']} onChange={this.handleChange} disabled={this.props.formDisable} />
        )
    }
}

const mapToStateToProps = (state) => {
    return {
        fieldsAsProps: state.fields,
        formDisable: state.formDisable
    }
}

const mapDispatchToProps = dispatch => {
    return {
      updateForm: (value) => dispatch(updateForm(value))
    }
}

export default connect(
  mapToStateToProps, 
  mapDispatchToProps
)(FormField);