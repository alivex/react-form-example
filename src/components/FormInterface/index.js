import * as React from 'react';

import { Col, FormGroup, Label } from 'reactstrap';
import FormInput from '../FormInput';

export class FormInterface extends React.Component{

  render(){
      const data = { 
          name: this.props.inputs.name, 
          type: this.props.inputs.type, 
          placeholder: this.props.inputs.placeholder
      }
      return(
          <FormGroup row={true}>
            <Label for={ `${this.props.inputs.name}` } sm={2}>{ this.props.inputs.label }</Label>
            <Col sm={10}>
              <FormInput data={ data }/>
            </Col>
          </FormGroup>
      )
  }
}

export default FormInterface;