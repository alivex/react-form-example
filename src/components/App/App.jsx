import * as React from 'react';
import { connect } from 'react-redux';

import FormInterface from '../FormInterface';
import Loader from '../Loader';
import { Container, Form, Button } from 'reactstrap';
import { submitForm, submitFormSuccess } from '../../store/actions'

import { submitTimeout } from '../../submitMock';

import './custom.scss';
import './App.css';

class App extends React.Component {

  // creating form ref to reset inputs values
  formRef = React.createRef();
  
  // Submit form
  submitForm(e) {
    
    e.preventDefault();

    this.props.submitForm(true);
    
    // Setting the timeout for 5000 ms
    submitTimeout(5000, 'Submit Form')
    .then((value) => {
      this.props.submitFormSuccess(false);
      this.formRef.current.reset()
    });
    
  }

  render() {
    return (
      <Container className="App">

        <Loader open={ this.props.formDisable } />

        <h2>Form</h2>
        <Form className="form" onSubmit={ (e) => this.submitForm(e) } innerRef={ this.formRef }>
          {this.props.form.map((input, index) => (
            <FormInterface 
              key={index}
              inputs={input}
            />
          ))}
          <Button className="submitBtn" disabled={ this.props.formDisable }>Submit</Button>
        </Form>
        
      </Container>
    );
  }
}

const mapToStateToProps = state => {
    return {
        form: state.form,
        formDisable: state.formDisable
    }
};

const mapDispatchToProps = dispatch => {
    return {
        submitForm: (value) => dispatch(submitForm(value)),
        submitFormSuccess: (value) => dispatch(submitFormSuccess(value)),
    };
};

export default connect(
    mapToStateToProps,
    mapDispatchToProps
)(App);
