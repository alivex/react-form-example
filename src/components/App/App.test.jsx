import  App from './App';
import Loader from '../Loader';
import { Form } from 'reactstrap';
import { FormInterface } from '../FormInterface';
import FormInput from '../FormInput';

import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

import AppState from '../../store/AppState';

const mockStore = configureMockStore();
const store = mockStore(AppState);

    it("renders form component", () => {
      const wrapper = render(
        <Form className="form" onSubmit={() => {}} />
      );
      expect(wrapper).toMatchSnapshot();
    });
    
    it("Loader opens correctly when form is disabled", () => {
      const wrapper = mount(
        <Loader open={true} />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should call submitForm on form submission', () => {
        const submitSpy = sinon.spy();
        const wrapper = mount(
          <Form className="form" onSubmit={ submitSpy } />
        );
        wrapper
          .simulate("submit");
      expect(submitSpy.calledOnce).toBe(true);
      
    });
    
    describe('Form Component', () => {
      
      const mockSubmitfn = jest.fn();
      let wrapper;
      
     
      beforeEach(() => {

        wrapper = mount(
          <Provider store={store}>
            <App onSubmit={ mockSubmitfn } />
          </Provider>
        );

      })

      it('should be called with the name, email, and message in the state as arguments', () => {
        
        const input = wrapper.find('.form-control');
        
         // fill in name field with Andrew Jacksson
          input.at(0).simulate(
           'change', 
           {target: 
             {name: 'nameInp', value: 'Andrew Jacksson'}
           }
         )
        

         // fill in email field with andrew@jack.sson  
          input.at(1).simulate(
           'change', 
           {target: 
              {name: 'emailInp', value: 'andrew@jack.sson'}
           }
         )
         
         // fill in message field with My message is Hello  
          input.at(2).simulate(
           'change', 
           {target: 
              {name: 'messageInp', value: 'My message is Hello'}
           }
         )

         // simulate form submission   
         wrapper.simulate(
           'submit', 
           {preventDefault() {}}
         )
          
         // test to see store state has been updated
         const state = store.getState('fields');
         
         expect(state['fields']).toEqual(
           {messageInp: 'My message is Hello', emailInp: 'andrew@jack.sson', nameInp: 'Andrew Jacksson'}
         )
      })
    });
    