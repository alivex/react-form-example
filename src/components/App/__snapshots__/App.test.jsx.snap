// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader opens correctly when form is disabled 1`] = `
<Portal
  containerInfo={
    <div
      style="position: relative; z-index: 1050;"
      tabindex="-1"
    >
      <div
        class=""
      >
        <div
          class="modal fade"
          role="dialog"
          style="display: block;"
          tabindex="-1"
        >
          <div
            class="modal-dialog modal-dialog-centered"
            role="document"
          >
            <div
              class="modal-content"
            >
              <div
                class="modal-body"
              >
                <div
                  class="spinner-grow text-primary"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-dark"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-secondary"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-info"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-warning"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-success"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-danger"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-success"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-danger"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-warning"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-info"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-secondary"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-dark"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-primary"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="modal-backdrop fade"
        />
      </div>
    </div>
  }
>
  <div
    className=""
  >
    <div
      className="modal fade"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      role="dialog"
      style={
        Object {
          "display": "block",
        }
      }
      tabIndex="-1"
    >
      <div
        className="modal-dialog modal-dialog-centered"
        role="document"
      >
        <div
          className="modal-content"
        >
          <div
            className="modal-body"
          >
            <div
              className="spinner-grow text-primary"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-dark"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-secondary"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-info"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-warning"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-success"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-danger"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-success"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-danger"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-warning"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-info"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-secondary"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-dark"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-primary"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="modal-backdrop fade"
    />
  </div>
</Portal>
`;

exports[`Loader opens correctly when submit 1`] = `
<Portal
  containerInfo={
    <div
      style="position: relative; z-index: 1050;"
      tabindex="-1"
    >
      <div
        class=""
      >
        <div
          class="modal fade"
          role="dialog"
          style="display: block;"
          tabindex="-1"
        >
          <div
            class="modal-dialog modal-dialog-centered"
            role="document"
          >
            <div
              class="modal-content"
            >
              <div
                class="modal-body"
              >
                <div
                  class="spinner-grow text-primary"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-dark"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-secondary"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-info"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-warning"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-success"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-danger"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-success"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-danger"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-warning"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-info"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-secondary"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-grow text-dark"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
                <div
                  class="spinner-border text-primary"
                  role="status"
                >
                  <span
                    class="sr-only"
                  >
                    Loading...
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="modal-backdrop fade"
        />
      </div>
    </div>
  }
>
  <div
    className=""
  >
    <div
      className="modal fade"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      role="dialog"
      style={
        Object {
          "display": "block",
        }
      }
      tabIndex="-1"
    >
      <div
        className="modal-dialog modal-dialog-centered"
        role="document"
      >
        <div
          className="modal-content"
        >
          <div
            className="modal-body"
          >
            <div
              className="spinner-grow text-primary"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-dark"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-secondary"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-info"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-warning"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-success"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-danger"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-success"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-danger"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-warning"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-info"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-secondary"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-grow text-dark"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
            <div
              className="spinner-border text-primary"
              role="status"
            >
              <span
                className="sr-only"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="modal-backdrop fade"
    />
  </div>
</Portal>
`;

exports[`renders component 1`] = `
<form
  class="form"
/>
`;

exports[`renders form component 1`] = `
<form
  class="form"
/>
`;
